CMakeLists.txt

```diff
@@ -225,10 +225,10 @@ endif()
 
 find_package(libyuv QUIET) # not required
 if(libyuv_FOUND)
-    # libyuv 1755 exposed all of the I*Matrix() functions, which libavif relies on.
+    # libyuv 1778 exposed ScalePlane_12 function, which libavif relies on.
     if(NOT LIBYUV_VERSION)
         message(STATUS "libavif: libyuv found, but version unknown; libyuv-based fast paths disabled.")
-    elseif(LIBYUV_VERSION LESS 1755)
+    elseif(LIBYUV_VERSION LESS 1778)
```

Comment:
Hi Bruno,

I just realized that this approach has a serious downside: it will cause libavif to not use libyuv for RGB-to-YUV and YUV-to-RGB conversions if libyuv is not version 1778 or newer.

libavif uses libyuv for two purposes: RGB<->YUV conversion and scaling. Between the two, RGB<->YUV conversion is much more important.

I suggest we use a different approach. We check for libyuv version at the place where we call ScalePlane_12(). Please search for "LIBYUV_VERSION" in libavif/src/reformat_libyuv.c for examples.

I think something like this patch should work. In this patch I use libyuv version 1774 for simplicity and leave the responsibility of using a libyuv version that is good enough to the user.

```
diff --git a/src/scale.c b/src/scale.c
index 6ca651b..89bf2fe 100644
--- a/src/scale.c
+++ b/src/scale.c
@@ -71,6 +71,13 @@ avifBool avifImageScale(avifImage * image, uint32_t dstWidth, uint32_t dstHeight
     image->height = dstHeight;
 
     if (srcYUVPlanes[0] || srcAlphaPlane) {
+#if LIBYUV_VERSION < 1774
+        if (image->depth > 8) {
+            avifDiagnosticsPrintf(diag, "ScalePlane_12() is not available in libyuv version %d", LIBYUV_VERSION);
+            return AVIF_FALSE;
+        }
+#endif
+
         // A simple conservative check to avoid integer overflows in libyuv's ScalePlane() and
         // ScalePlane_12() functions.
         if (srcWidth > 16384) {
@@ -103,11 +110,13 @@ avifBool avifImageScale(avifImage * image, uint32_t dstWidth, uint32_t dstHeight
             const uint32_t dstW = (i == AVIF_CHAN_Y) ? dstWidth : dstUVWidth;
             const uint32_t dstH = (i == AVIF_CHAN_Y) ? dstHeight : dstUVHeight;
             if (image->depth > 8) {
+#if LIBYUV_VERSION >= 1774
                 uint16_t * const srcPlane = (uint16_t *)srcYUVPlanes[i];
                 const uint32_t srcStride = srcYUVRowBytes[i] / 2;
                 uint16_t * const dstPlane = (uint16_t *)image->yuvPlanes[i];
                 const uint32_t dstStride = image->yuvRowBytes[i] / 2;
                 ScalePlane_12(srcPlane, srcStride, srcW, srcH, dstPlane, dstStride, dstW, dstH, AVIF_LIBYUV_FILTER_MODE);
+#endif
             } else {
                 uint8_t * const srcPlane = srcYUVPlanes[i];
                 const uint32_t srcStride = srcYUVRowBytes[i];
@@ -126,11 +135,13 @@ avifBool avifImageScale(avifImage * image, uint32_t dstWidth, uint32_t dstHeight
         avifImageAllocatePlanes(image, AVIF_PLANES_A);
 
         if (image->depth > 8) {
+#if LIBYUV_VERSION >= 1774
             uint16_t * const srcPlane = (uint16_t *)srcAlphaPlane;
             const uint32_t srcStride = srcAlphaRowBytes / 2;
             uint16_t * const dstPlane = (uint16_t *)image->alphaPlane;
             const uint32_t dstStride = image->alphaRowBytes / 2;
             ScalePlane_12(srcPlane, srcStride, srcWidth, srcHeight, dstPlane, dstStride, dstWidth, dstHeight, AVIF_LIBYUV_FILTER_MODE);
+#endif
         } else {
             uint8_t * const srcPlane = srcAlphaPlane;
             const uint32_t srcStride = srcAlphaRowBytes;
```

Reply:
Alternatively, we can use `ScalePlane_16` instead if `LIBYUV_VERSION < 1774`.
`ScalePlane_12` has special optimization for 2x scale up using linear/bilinear filter (when scaling up, box and bilinear are the same), and for the rest cases, it simply calls `ScalePlane_16`, which handles 2x scale up just fine (slower though).

Reply:
Seems a good idea. However I’m not exactly sure whether the arguments for `ScalePlane_16` are the same as for the `_12` versions, and especially I don’t know (the) code enough to avoid issues such as https://github.com/AOMediaCodec/libavif/issues/715.

Reply:
The arguments are the same: https://chromium.googlesource.com/libyuv/libyuv/+/a04e4f87fbf40405707b1d0ae9fcba8fc93f7856/source/scale.cc#2073, so it should be fine to use `ScalePlane_16` when `ScalePlane_12` is unavailable.